File: src/adapters/anthropic/to-ir.ts
import type {
  IrContent,
  IrMessage,
  IrRequest,
  IrRole,
  IrTool,
  ProviderExtensions,
} from '../../ir/types.js';
import {
  MODELLED_MESSAGE_KEYS,
  MODELLED_REQUEST_KEYS,
  MODELLED_TEXT_KEYS,
  MODELLED_TOOL_RESULT_KEYS,
  MODELLED_TOOL_USE_KEYS,
  extractPassthrough,
  isPlainObject,
} from './passthrough.js';

/** An Anthropic `/v1/messages` request body, unvalidated. */
export type AnthropicRequestBody = Record<string, unknown>;

function toOpaque(raw: unknown): IrContent {
  return { kind: 'opaque', raw };
}

function toTextContent(block: Record<string, unknown>): IrContent {
  const content: IrContent = {
    kind: 'text',
    text: String(block['text'] ?? ''),
    compressible: true,
  };
  if ('cache_control' in block) content.cacheControl = block['cache_control'];
  const rest = extractPassthrough(block, MODELLED_TEXT_KEYS);
  if (rest !== undefined) content.passthrough = rest;
  return content;
}

function toToolUseContent(block: Record<string, unknown>): IrContent {
  const content: IrContent = {
    kind: 'tool_use',
    id: String(block['id'] ?? ''),
    name: String(block['name'] ?? ''),
    input: block['input'],
  };
  if ('cache_control' in block) content.cacheControl = block['cache_control'];
  const rest = extractPassthrough(block, MODELLED_TOOL_USE_KEYS);
  if (rest !== undefined) content.passthrough = rest;
  return content;
}

function toToolResultContent(block: Record<string, unknown>): IrContent {
  const raw = block['content'];
  const isContentString = typeof raw === 'string';
  const content: IrContent = {
    kind: 'tool_result',
    toolUseId: String(block['tool_use_id'] ?? ''),
    content: toContentArray(raw),
    isContentString,
  };
  if ('is_error' in block) content.isError = block['is_error'] === true;
  if ('cache_control' in block) content.cacheControl = block['cache_control'];
  const rest = extractPassthrough(block, MODELLED_TOOL_RESULT_KEYS);
  if (rest !== undefined) content.passthrough = rest;
  return content;
}

/**
 * `thinking` and `redacted_thinking` carry a signature the API validates, so
 * they are kept as raw wire values rather than being decomposed.
 */
const THINKING_TYPES: readonly string[] = ['thinking', 'redacted_thinking'];

function toContentBlock(block: unknown): IrContent {
  if (!isPlainObject(block)) return toOpaque(block);
  const type = block['type'];
  if (type === 'text') return toTextContent(block);
  if (type === 'tool_use') return toToolUseContent(block);
  if (type === 'tool_result') return toToolResultContent(block);
  if (typeof type === 'string' && THINKING_TYPES.includes(type)) {
    return { kind: 'thinking', raw: block };
  }
  return toOpaque(block);
}

/**
 * Normalizes the string-or-array content forms to an array. The original form
 * is recorded separately so `fromIR` can reproduce it.
 */
function toContentArray(raw: unknown): IrContent[] {
  if (raw === undefined || raw === null) return [];
  if (typeof raw === 'string') return [{ kind: 'text', text: raw, compressible: true }];
  if (!Array.isArray(raw)) return [toOpaque(raw)];
  return raw.map(toContentBlock);
}

function toMessage(raw: unknown): IrMessage {
  if (!isPlainObject(raw)) {
    return { role: 'user', content: [toOpaque(raw)], isContentString: false };
  }
  const content = raw['content'];
  const message: IrMessage = {
    role: raw['role'] as IrRole,
    content: toContentArray(content),
    isContentString: typeof content === 'string',
  };
  const rest = extractPassthrough(raw, MODELLED_MESSAGE_KEYS);
  if (rest !== undefined) message.passthrough = rest;
  return message;
}

/**
 * Only a string or block-array `system` is modelled; any other form (including
 * an explicit null) stays in passthrough so it round-trips as written.
 */
function isSystemModelled(body: AnthropicRequestBody): boolean {
  const system = body['system'];
  return typeof system === 'string' || Array.isArray(system);
}

function toSystem(raw: unknown): { system: IrContent[] | null; isSystemString: boolean } {
  if (typeof raw !== 'string' && !Array.isArray(raw)) {
    return { system: null, isSystemString: false };
  }
  return { system: toContentArray(raw), isSystemString: typeof raw === 'string' };
}

function toTools(raw: unknown): IrTool[] {
  if (!Array.isArray(raw)) return [];
  return raw.map((tool) => ({ raw: tool }));
}

/**
 * An empty `tools: []` is indistinguishable from an absent one once modelled as
 * an array, so it stays in passthrough and only a populated list is modelled.
 */
function isToolsModelled(body: AnthropicRequestBody): boolean {
  const tools = body['tools'];
  return Array.isArray(tools) && tools.length > 0;
}

function modelledRequestKeys(body: AnthropicRequestBody): readonly string[] {
  const dropped = new Set<string>();
  if (!isToolsModelled(body)) dropped.add('tools');
  if (!isSystemModelled(body)) dropped.add('system');
  return MODELLED_REQUEST_KEYS.filter((key) => !dropped.has(key));
}

/**
 * Converts an Anthropic request body into the IR. Unknown fields and unknown
 * block types are preserved rather than dropped, so an unrecognized future
 * shape degrades to passthrough instead of data loss.
 */
export function toIR(body: AnthropicRequestBody): IrRequest {
  const { system, isSystemString } = toSystem(body['system']);
  const messages = Array.isArray(body['messages']) ? body['messages'] : [];
  const extensions: ProviderExtensions = {};
  if (isSystemString) extensions.isSystemString = true;
  return {
    model: String(body['model'] ?? ''),
    maxTokens: Number(body['max_tokens'] ?? 0),
    system,
    messages: messages.map(toMessage),
    tools: isToolsModelled(body) ? toTools(body['tools']) : [],
    extensions,
    passthrough: extractPassthrough(body, modelledRequestKeys(body)) ?? {},
  };
}

File: internal/adapters/anthropic/toir.go
package anthropic

import (
	"math"
	"strconv"
	"strings"

	"github.com/carlelieser/caveman/internal/ir"
)

// thinking and redacted_thinking carry a signature the API validates, so they
// are kept as raw wire values rather than being decomposed.
var thinkingTypes = []string{"thinking", "redacted_thinking"}

// jsString reproduces JavaScript's `String(value ?? ”)`. Absent and null both
// reach the empty string; everything else takes the coercion the language
// would apply, so a body carrying the wrong type still round-trips as one.
func jsString(value ir.Value, present bool) string {
	if !present {
		return ""
	}
	switch typed := value.(type) {
	case nil, ir.Null:
		return ""
	case ir.String:
		return string(typed)
	case ir.Bool:
		if typed {
			return "true"
		}
		return "false"
	case ir.Number:
		return typed.Literal()
	case ir.Array:
		parts := make([]string, len(typed))
		for i, item := range typed {
			if _, isNull := item.(ir.Null); isNull || item == nil {
				parts[i] = ""
				continue
			}
			parts[i] = jsString(item, true)
		}
		return strings.Join(parts, ",")
	case *ir.Object:
		return "[object Object]"
	}
	return ""
}

// jsNumber reproduces JavaScript's `Number(value ?? 0)`, keeping the literal
// the wire carried when the value was already a number so re-serializing it
// cannot change the bytes.
func jsNumber(value ir.Value, present bool) ir.Number {
	if !present {
		return ir.NumberFromInt(0)
	}
	switch typed := value.(type) {
	case nil, ir.Null:
		return ir.NumberFromInt(0)
	case ir.Number:
		return typed
	case ir.Bool:
		if typed {
			return ir.NumberFromInt(1)
		}
		return ir.NumberFromInt(0)
	case ir.String:
		return numberFromString(string(typed))
	case ir.Array:
		if len(typed) == 0 {
			return ir.NumberFromInt(0)
		}
		if len(typed) == 1 {
			return jsNumber(typed[0], true)
		}
	}
	return ir.NumberFromFloat(math.NaN())
}

func numberFromString(text string) ir.Number {
	trimmed := strings.TrimSpace(text)
	if trimmed == "" {
		return ir.NumberFromInt(0)
	}
	parsed, err := strconv.ParseFloat(trimmed, 64)
	if err != nil {
		return ir.NumberFromFloat(math.NaN())
	}
	return ir.NumberFromFloat(parsed)
}

func toOpaque(raw ir.Value) ir.Content {
	return &ir.OpaqueContent{Raw: raw}
}

func toTextContent(block *ir.Object) ir.Content {
	text, present := block.Get("text")
	content := &ir.TextContent{
		Text:     jsString(text, present),
		KeyOrder: ir.KeyOrder(block.Keys()),
	}
	if cacheControl, ok := block.Get("cache_control"); ok {
		content.CacheControl = cacheControl
	}
	content.Passthrough = extractPassthrough(block, modelledTextKeys)
	return content
}

func toToolUseContent(block *ir.Object) ir.Content {
	id, hasID := block.Get("id")
	name, hasName := block.Get("name")
	input, hasInput := block.Get("input")
	content := &ir.ToolUseContent{
		ID:       jsString(id, hasID),
		Name:     jsString(name, hasName),
		Input:    input,
		HasInput: hasInput,
		KeyOrder: ir.KeyOrder(block.Keys()),
	}
	if cacheControl, ok := block.Get("cache_control"); ok {
		content.CacheControl = cacheControl
	}
	content.Passthrough = extractPassthrough(block, modelledToolUseKeys)
	return content
}

func toToolResultContent(block *ir.Object) ir.Content {
	raw, _ := block.Get("content")
	_, isContentString := asString(raw)
	toolUseID, hasToolUseID := block.Get("tool_use_id")
	content := &ir.ToolResultContent{
		ToolUseID:       jsString(toolUseID, hasToolUseID),
		Content:         toContentArray(raw),
		IsContentString: isContentString,
		KeyOrder:        ir.KeyOrder(block.Keys()),
	}
	if isError, ok := block.Get("is_error"); ok {
		flag := isError == ir.Bool(true)
		content.IsError = &flag
	}
	if cacheControl, ok := block.Get("cache_control"); ok {
		content.CacheControl = cacheControl
	}
	content.Passthrough = extractPassthrough(block, modelledToolResultKeys)
	return content
}

func toContentBlock(block ir.Value) ir.Content {
	object, ok := asObject(block)
	if !ok {
		return toOpaque(block)
	}
	rawType, _ := object.Get("type")
	blockType, isString := asString(rawType)
	if !isString {
		return toOpaque(block)
	}
	switch blockType {
	case "text":
		return toTextContent(object)
	case "tool_use":
		return toToolUseContent(object)
	case "tool_result":
		return toToolResultContent(object)
	}
	if contains(thinkingTypes, blockType) {
		return &ir.ThinkingContent{Raw: block}
	}
	return toOpaque(block)
}

// toContentArray normalizes the string-or-array content forms to an array. The
// original form is recorded separately so FromIR can reproduce it.
func toContentArray(raw ir.Value) []ir.Content {
	if raw == nil {
		return []ir.Content{}
	}
	if _, isNull := raw.(ir.Null); isNull {
		return []ir.Content{}
	}
	if text, ok := asString(raw); ok {
		return []ir.Content{&ir.TextContent{Text: text}}
	}
	items, ok := asArray(raw)
	if !ok {
		return []ir.Content{toOpaque(raw)}
	}
	blocks := make([]ir.Content, len(items))
	for i, item := range items {
		blocks[i] = toContentBlock(item)
	}
	return blocks
}

func toMessage(raw ir.Value) ir.Message {
	object, ok := asObject(raw)
	if !ok {
		return ir.Message{Role: ir.RoleUser, Content: []ir.Content{toOpaque(raw)}}
	}
	content, _ := object.Get("content")
	_, isContentString := asString(content)
	role, _ := object.Get("role")
	roleText, roleIsString := asString(role)
	message := ir.Message{
		Role:            ir.Role(roleText),
		Content:         toContentArray(content),
		IsContentString: isContentString,
		KeyOrder:        ir.KeyOrder(object.Keys()),
	}
	// The IR types role as a string. A wire body carrying anything else keeps
	// its value in passthrough rather than losing it to the coercion.
	modelled := modelledMessageKeys
	if !roleIsString {
		modelled = []string{"content"}
	}
	message.Passthrough = extractPassthrough(object, modelled)
	return message
}

// Only a string or block-array `system` is modelled; any other form (including
// an explicit null) stays in passthrough so it round-trips as written.
func isSystemModelled(body *ir.Object) bool {
	system, _ := body.Get("system")
	if _, ok := asString(system); ok {
		return true
	}
	_, ok := asArray(system)
	return ok
}

func toSystem(raw ir.Value) (system []ir.Content, hasSystem bool, isSystemString bool) {
	if _, ok := asString(raw); !ok {
		if _, isArray := asArray(raw); !isArray {
			return nil, false, false
		}
	}
	_, isString := asString(raw)
	return toContentArray(raw), true, isString
}

func toTools(raw ir.Value) []ir.Tool {
	items, ok := asArray(raw)
	if !ok {
		return []ir.Tool{}
	}
	tools := make([]ir.Tool, len(items))
	for i, item := range items {
		tools[i] = ir.Tool{Raw: item}
	}
	return tools
}

// An empty `tools: []` is indistinguishable from an absent one once modelled as
// a slice, so it stays in passthrough and only a populated list is modelled.
func isToolsModelled(body *ir.Object) bool {
	tools, _ := body.Get("tools")
	items, ok := asArray(tools)
	return ok && len(items) > 0
}

func requestKeysToModel(body *ir.Object) []string {
	kept := make([]string, 0, len(modelledRequestKeys))
	for _, key := range modelledRequestKeys {
		if key == "tools" && !isToolsModelled(body) {
			continue
		}
		if key == "system" && !isSystemModelled(body) {
			continue
		}
		kept = append(kept, key)
	}
	return kept
}

// ToIR converts an Anthropic request body into the IR. Unknown fields and
// unknown block types are preserved rather than dropped, so an unrecognized
// future shape degrades to passthrough instead of data loss.
func ToIR(body *ir.Object) ir.Request {
	rawSystem, _ := body.Get("system")
	system, hasSystem, isSystemString := toSystem(rawSystem)

	rawMessages, _ := body.Get("messages")
	messageItems, _ := asArray(rawMessages)
	messages := make([]ir.Message, len(messageItems))
	for i, item := range messageItems {
		messages[i] = toMessage(item)
	}

	model, hasModel := body.Get("model")
	maxTokens, hasMaxTokens := body.Get("max_tokens")

	tools := []ir.Tool{}
	if isToolsModelled(body) {
		rawTools, _ := body.Get("tools")
		tools = toTools(rawTools)
	}

	passthrough := extractPassthrough(body, requestKeysToModel(body))
	if passthrough == nil {
		passthrough = ir.NewObject()
	}

	return ir.Request{
		Model:     jsString(model, hasModel),
		MaxTokens: jsNumber(maxTokens, hasMaxTokens),
		System:    system,
		HasSystem: hasSystem,
		Messages:  messages,
		Tools:     tools,
		Extensions: ir.Extensions{
			IsSystemString: isSystemString,
			KeyOrder:       ir.KeyOrder(body.Keys()),
		},
		Passthrough: passthrough,
	}
}
